test/metabase/lib/card_test.cljc

```diff
@@ -91,3 +94,62 @@
              (mapv :name (get-in lib.tu/mock-cards [:orders :result-metadata]))))
       (is (= ["ID" "SUBTOTAL" "TOTAL" "TAX" "DISCOUNT" "QUANTITY" "CREATED_AT" "PRODUCT_ID" "USER_ID"]
              (mapv :name (lib/visible-columns venues-query)))))))
+
+(deftest ^:parallel returned-columns-31769-test
+  (testing "Cards with joins should return correct column metadata/refs (#31769)"
+    (let [metadata-provider (lib.tu.mocks-31769/mock-metadata-provider meta/metadata-provider meta/id 1 2)
+          card              (lib.metadata/card metadata-provider 1)
+          q                 (:dataset-query card)
+          cols              (lib/returned-columns q)]
+      (is (=? [{:name                         "CATEGORY"
+                :lib/source                   :source/breakouts
+                :lib/source-column-alias      "CATEGORY"
+                :metabase.lib.join/join-alias "Products"
+                :lib/desired-column-alias     "Products__CATEGORY"}
+               {:name                     "count"
+                :lib/source               :source/aggregations
+                :lib/source-column-alias  "count"
+                :lib/desired-column-alias "count"}]
+              cols))
+      (is (=? [[:field {:join-alias "Products"} (meta/id :products :category)]
+               [:aggregation {} string?]]
+              (map lib.ref/ref cols))))))
```

Comment:
Isn't it a bad sign if `lib/returned-columns` is required to provide information about the internals of a query? A native query wouldn't.
I'm confused about the semantics of `returned-columns`. I thought it would describe what columns I get if I run the given query. And then I would expect, for example, `:name` to be `"Products__CATEGORY"`.

Reply:
Yeah in hindsight maybe `returned-columns` wasn't the clearest name, maybe we should have just kept the original name `expected-columns` from https://github.com/metabase/metabase/pull/32028#discussion_r1248050935

`returned-columns` gives you metadata about the columns in a stage that will show up in the results (if this is the last stage) or be available to the next stage (if this is not the last stage). This metadata is still relative to the current stage, so calling `lib/ref` will return refs that are appropriate for use **in that same stage**. For example if `returned-columns` tells you you have a `category` column and you add an order by to the current stage using that info it would update the query like

```clj
{:order-by [[:asc {} [:field {:join-alias "Products"} 1]]]}
```

or whatever.

`returned-columns` is just returning metadata bout the stuff relative to that stage. `ref` is just generating refs based on that metadata. These refs are correct if you are going to use them in this stage.


If you added another stage, then the ref for that same column in the new stage would be something like

```clj
[:field {} "Products__CATEGORY"]
```

Reply:
Are you saying that `returned-columns` is a subset of `visible-columns`?

Reply:
Yes

Reply:
Then I would expect an implementation of `returned-columns` that just filters `visible-columns`. What we have now looks like two parallel implementations. The invariant that one is a subset of the other is not obvious.

Reply:
The default implementation of `visible-columns` **is** `returned-columns`, custom implementations just add other stuff in addition. You're right tho that those implementations could be consolidated a bit more.